components/session-service/server/server.go

```diff
@@ -675,7 +680,97 @@ func (s *Server) refreshHandler(w http.ResponseWriter, r *http.Request) {
 		IDToken string `json:"id_token"`
 	}{rawIDToken}
 	if err := json.NewEncoder(w).Encode(returnData); err != nil {
-		http.Error(w, errors.Wrap(err, "failed to set marshal id_token").Error(), http.StatusInternalServerError)
+		JSONError(w, prepareError(http.StatusInternalServerError, errors.Wrap(err, "failed to set marshal id_token").Error()), http.StatusInternalServerError)
+		return
+	}
+}
+
+type RefreshToken struct {
+	RefreshToken string `json:"refresh_token"`
+	GrantType    string `json:"grant_type"`
+}
+
+type ErrorResponse struct {
+	ErrorCode int    `json:"error_code"`
+	ErrorDesc string `json:"error"`
+}
+
+func prepareError(code int, errMsg string) ErrorResponse {
+	return ErrorResponse{
+		ErrorCode: code,
+		ErrorDesc: errMsg,
+	}
+}
+
+func JSONError(w http.ResponseWriter, err interface{}, code int) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+	w.WriteHeader(code)
+	json.NewEncoder(w).Encode(err)
+}
+
+func (s *Server) refreshApiHandler(w http.ResponseWriter, r *http.Request) {
+	var data RefreshToken
+	// Try to decode the request body into the struct. If there is an error,
+	// respond to the client with the error message and a 400 status code.
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		JSONError(w, prepareError(http.StatusBadRequest, err.Error()), http.StatusBadRequest)
+		return
+	}
+	if data.RefreshToken == "" || data.GrantType == "" {
+		JSONError(w, prepareError(http.StatusBadRequest, "Send All Input Parameters"), http.StatusBadRequest)
+		return
+	}
+	refreshToken := data.RefreshToken
+	idToken, err := util.ExtractBearerToken(r)
+	if err != nil {
+		s.log.Debug("no bearer token")
+		JSONError(w, prepareError(http.StatusUnauthorized, err.Error()), http.StatusUnauthorized)
+		return
+	}
+
+	isBlacklisted, err := s.idTokenBlackLister.IsIdTokenBlacklisted(idToken)
+	if err != nil {
+		JSONError(w, prepareError(http.StatusInternalServerError, err.Error()), http.StatusInternalServerError)
+		return
+	}
+	if isBlacklisted {
+		JSONError(w, prepareError(http.StatusUnauthorized, "ID Token is Expired"), http.StatusUnauthorized)
+		return
+	}
+
+	resp, err := s.client.RefreshTokenValidator(refreshToken)
+	if err != nil {
+		JSONError(w, prepareError(http.StatusInternalServerError, err.Error()), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
```

Comment:
Is this check needed? Could we handle this in err?

Reply:
Yes, it is needed.
As it's a http handler we can't return the error. we need to set the status code then only we can check it if something is wrong there.